### ISSUE TITLE
simple POJO API

### ISSUE THREAD

**Issue by ghost** [makumba/makumba#776]
**Reported by @cristianbogdan on 18 May 2010 10:18 UTC**
today mak:form creates a hierarchical Dictionary (i.e. dic  with fields from multiple, inter-related objects) the present database API 

the idea of this entry is to preserve this functionality while working with POJOs. one important issue here is that the makumba POJO classes are automatically generated (bug 1229), programmers don't mess up with them and therefore we have full control on how to implement the setter methods, etc. this makes things much easier than for a persistence manager.

so the suggestion here would be to do the minimal work necessary to generate a _graph_ of interrelated POJOs (one root object and its related fields) from the hierarchical dictionary passed by forms via the HTTP query string

the old Database API can stay (as the hierarchical dictionary -> POJO graph is needed anyway for forms) but deprecate

in the long term, we can pass the  POJO graph (and the create/update/delete calls) to JPA or any persistence manager. we can already prepare in terms of CRD API compatibility/similarity.

in the short term, we can pass that graph to our own database API. we already have the code that takes the hierarchical Dictionary and transforms it in plain (single-type) Dictionary, ready for the low-level database insert or update. so the idea is simply to generate the plain POJOs instead of the plain Dictionaries and write them to the db in the right order. i don't think this is so hard.

one difference from persistence managers is that we don't need to retrieve the whole object before updating it. like 

s= retrieve("best.Student", primaryKey);
s.setName("bla");
update(s);

the retrieve function simply creates an empty best.Student with that primary key.
the setName sets the name but also stores that name is dirty, so update() will update only that field.

if a programmer in an onEdit() BL method wants to do s.getSurname() (i.e. a yet-unknown field), that will trigger an object read based on the primary key. but the question is whether we want to provide such an API or provide just Dictionary, and use JPA for anything POJO-like, as maybe we don't want to recreate all JPA functionality, just prepare more JPA compatibility

but it may be easy to provide object retrieval and related-object retrieval in a low-performance way and that may be handy in BL.

so we definitely should consider BL methods here.

Migrated-From: http://trac.makumba.org/ticket/1231

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 18 May 2010 10:39 UTC**
(In reply to bug 1226 comment #15)
> I'm a bit skeptical of going down that road because that sounds awfully a lot
> like implementing a ORM solution. I mean for read-only it is fine. but as soon
> (and this will probably happen very soon) as someone wants to modify an object
> state you have to track this somehow. this is what all the idiomatic java is
> about. they use deep and sophisticated introspection methods to figure out what
> changed in the state of an object.

i agree on read-only and not automatic much detection of dirty objects. the objective here is a transformation from hierarchical dictionary to POJO graph and from Dictionary APIs to Object APIs with minimal effort possible!

> the above Bungee jump() changes the state of the student object, and that could
> be in many sophisticated ways such as
> student.getHormonalSystem().setAdrenalineLevel(1000);

yes, in that case the programmer is responsible to save the student POJO graph.

or actually, as long as he does changes only for _student_ makumba will save the POJO graph for him anyway (because it's onEdit(best.Student)! but if he changes other root objects, makumba will not detect that. 

> And perhaps there are ways of implementing it - but I am afraid that this is
> hard. I mean the hibernate folks spent years to make their idiomatic engine
> good enough, I don't think this should be our main focus.

hibernate does much more. it does automatic dirty objects detection, it does lazy fetching, it does caching, etc. we do none of that. 

> I know the problem of karamba and speed of loading. Before going into the
> design of a makumba-ORM-layer though I would:

well if it turns out that there is a suitable JPA implementation that BEST can adopt, i have nothing against. but the hierarchical Dictionary-> POJO graph transformation has to be done in any case.

for the onEdit(best.Student s) API, if we deem we can't support it, we don't support it, we can afford to do that as it's a makumba-specific API.

but i also sense a golden opportunity for us to reuse our great code and provide a super lightweight JPA implementation, with a decent query language (MQL) and a simple, incremental schema update.  that would be rather nice.

> - investigate how JPA manages schema update, if they have the incremental one
> - make a speed test with K! (also simply by updating the hibernate version,
> maybe they improved on that)
> - see http://www.informit.com/articles/article.aspx?p=353737 and search for
> "incremental", there seems to be a tool to upgrade just a set of types in the
> schema, that can be talked to programmatically.

this can be done in _parallel_ with things that we have to do anyway. like MDD->POJO class (bug 1229), query -> JPA query API (bug 1230), and hierarchical Dic-> POJO graph (bug 1231)
 
> What about looking into using JPA as a database layer and implement a
> dictionary -> JPA bridge. that way K! still uses dictionary in many places and
> can use the JPA POJOs in new places.

well forms use hierarchical dictionary anyway, hence this entry.

**Comment by ghost** [comment]
**Comment by @cristianbogdan on 18 May 2010 16:32 UTC**
about the hierarchical Dictionary -> POJO graph business. it may be that JSF already works at POJO graph level?

like they are aware of the POJOs (beans?) but also about the relations between them?

in that case mak:JSF will not need this. instead we could aim to ensure that the POJO graph we expect is similar to the one JSF works with

**Comment by ghost** [comment]
**Comment by @manuelbernhardt on 18 May 2010 18:23 UTC**
(In reply to comment #2)
> about the hierarchical Dictionary -> POJO graph business. it may be that JSF
> already works at POJO graph level?
> 
> like they are aware of the POJOs (beans?) but also about the relations between
> them?
> 
> in that case mak:JSF will not need this. instead we could aim to ensure that
> the POJO graph we expect is similar to the one JSF works with

I am not sure what you mean by 'aware of'? In the sense that JSF forms would know how to populate #{myBean.some.thing}?

I think that is possible, but I am not sure. I need to check my JSF book which is at work. or you could check out the jsf playground app, create another bean next to the HelloWorld bean, make a reference in the helloWorld bean to that bean, and check on the JSF page if that works :)

(I am callibrating the battery of my mac so I can't play around)

In addition to this, there are very interesting things going on here:
http://seamframework.org/Documentation/JSF21 (the JBoss guys discussing stuff that should get into JSF 2.1). The design discussions shed a bit more light on JSF mechanisms and shortcomings (well, at least to me after reading the book).

**Comment by ghost** [comment]
**Comment by @manuelbernhardt on 19 May 2010 14:31 UTC**
I just made a test like this:

    <h:form id="person">
      <table>
        <tr>
          <th style="text-align: right;"><h:outputLabel for="email" value="Person name:" /></th>
          <td><h:inputText id="cname" value="#{helloWorld.person.name}" />
          <h:message for="name" errorClass="invalid" /></td>
        </tr>
        <tr>
          <th style="text-align: right;"><h:outputLabel for="email" value="Person name:" /></th>
          <td><h:inputText id="csurname" value="#{helloWorld.person.surname}" />
          <h:message for="surname" errorClass="invalid" /></td>
        </tr>
      </table>
      
      <p>
        <h:commandButton id="ccheck" action="check" value="Check values" />
        <h:outputText value=" All clear!" rendered="#{facesContext.postback and empty facesContext.messageList}" style="color: green;" />
      </p>
    </h:form>

in the HelloWorld bean I have


    private Person person = new Person();
    
    public Person getPerson() {
        return person;
    }

    public void setPerson(Person person) {
        this.person = person;
    }


and in the Person bean I added a System.out in the setters.

on execution (after pressing "Check") I get in the console:

baby your name is:manu
baby your surname is:el


so in short: yes, JSF knows how to traverse POJO graphs!

that is, as you see I need to give it an instance of Person() to traverse. Also I don't know yet how to bind a specific instance of a POJO to the JSF page but I guess that should be rather easy.

**Comment by ghost** [comment]
**Modified by @manuelbernhardt on 28 Jun 2010 07:22 UTC**

**Comment by ghost** [comment]
**Modified by @manuelbernhardt on 28 Jun 2010 07:24 UTC**

**Comment by ghost** [comment]
**Modified by @manuelbernhardt on 28 Jun 2010 07:36 UTC**
